awx/main/tasks/receptor.py

```diff
@@ -411,11 +413,16 @@ def _run_internal(self, receptor_ctl):
                     return
 
                 try:
-                    resultsock = receptor_ctl.get_work_results(self.unit_id, return_sockfile=True)
-                    lines = resultsock.readlines()
-                    receptor_output = b"".join(lines).decode()
-                    if receptor_output:
-                        self.task.runner_callback.delay_update(result_traceback=receptor_output)
+                    if state_name == 'Failed' and self.task.runner_callback.event_ct == 0:
+                        # if receptor work unit failed and no events were emitted, work results may
+                        # contain useful information about why the job failed. In case stdout is
+                        # massive, only ask for last 1000 bytes
+                        startpos = max(stdout_size - 1000, 0)
+                        resultsock = receptor_ctl.get_work_results(self.unit_id, startpos=startpos, return_sockfile=True)
+                        lines = resultsock.readlines()
```

Comment:
I'm worried about this hanging if the stream doesn't end with a `\n`. We may need to put the socket in non-blocking mode before doing this and just read the raw bytes.

Reply:
I did some local testing and readlines() seems to be okay with files that don't end in \n. If it gets an EOF it just stops there.

we could also just do read() or read(1000)

as for non-blocking I added `resultsock.setblocking(False)`. Hard to test this exactly, but I will make sure this codepath doesn't throw errors